compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp

```diff
@@ -85,6 +85,43 @@ struct HoistableTensorTypeInterface
   }
 };
 
+struct HoistableIndexTypeInterface
+    : public IREE::Util::HoistableTypeInterface::ExternalModel<
+          HoistableIndexTypeInterface, IndexType> {
+  bool isHoistableType(Type type) const {
+    return type.isIndex();
+  }
+  bool isHoistableLeafType(Type type) const {
+    return type.isIndex();
+  }
+  Type getPreferredStorageType(Type type) const {
+    return IntegerType::get(type.getContext(), 64);
+  }
+  static Value encodeStorageType(OpBuilder &builder, Location loc,
+                                 Type storageType, Value init) {
+    auto storageIndexType = dyn_cast<IntegerType>(storageType);
+    if (!storageIndexType || init.getType() == storageIndexType ||
+        !isa<IndexType>(init.getType())) {
+      return init;
+    }
+    return builder.create<IREE::Flow::TensorBitCastOp>(loc, storageIndexType,
```

Comment:
You should be using `arith.index_cast` instead of `Flow::TensorBitCast`, that might be where your problems are coming from.

Reply:
I changed it to arith.index_cast and got
```
error: 'util.initializer' op failed to inline into combined initializer
```
Seems the interface is not working.

Reply:
I remember hitting that error when turning off const-eval for the input IR you sent before, are you sure that this error is happening in the same pass?

Reply:
I think so. Also I would get the same error if I just add the code in JitGlobals.cpp without adding the interface.

Now the initializer seems changed
```
note: see current operation:
"util.initializer"() <{function_type = () -> ()}> ({
  %0 = "util.global.load"() <{global = @__hoisted_i64_36}> : () -> i64
  %1 = "arith.index_cast"(%0) : (i64) -> index
  %2 = "affine.apply"(%1) <{map = affine_map<()[s0] -> ((s0 + 16) * (s0 + 32))>}> : (index) -> index
  %3 = "arith.index_cast"(%2) : (index) -> i64
  "util.global.store"(%3) <{global = @__hoisted_i64_39}> : (i64) -> ()
  "util.return"() : () -> ()
}) {stream.affinity.default = #hal.device.affinity<@__device_0>} : () -> ()
    %139 = affine.apply #map16()[%109, %110, %134, %135, %136]
```
The same error I got before was on
```
note: see current operation:
"util.initializer"() <{function_type = () -> ()}> ({
  %0 = "util.global.load"() <{global = @__hoisted_index_8}> : () -> index
  %1 = "util.global.load"() <{global = @__hoisted_index_18}> : () -> index
  %2 = "util.global.load"() <{global = @__hoisted_index_17}> : () -> index
  %3 = "util.global.load"() <{global = @__hoisted_index}> : () -> index
  %4 = "affine.apply"(%2, %1, %3, %0) <{map = affine_map<()[s0, s1, s2, s3] -> (s0 + s1 + s2 + s3 + 64)>}> : (index, index, index, index) -> index
  "util.global.store"(%4) <{global = @__hoisted_index_21}> : (index) -> ()
  "util.return"() : () -> ()
}) {stream.affinity.default = #hal.device.affinity<@__device_0>} : () -> ()
    %111 = affine.apply #map9()[%107, %108, %99, %100]
```

Reply:
Can you add a test showing an example of hoisting for `index` so that we can verify whether the failure you're hitting is from the interface implementation or if it's an unrelated error?

Reply:
A lit test like https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Dialect/Util/Transforms/test/hoist_into_globals.mlir?

Reply:
The tests for the interface are here: https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/GlobalOptimization/test/hoist_into_globals.mlir because of the way that the interface registration is set up (to avoid a dep on upstream stuff in Util). Might be worth revisiting that so it isn't awkwardly split into two places.

Reply:
Added a test, seems it hoists the index as expected.
```
util.global private @__hoisted_i64 : i64
  util.initializer {
    %c0 = arith.constant 0 : index
    %0 = "iree_unregistered.const_expr"(%c0) : (index) -> index
    %1 = arith.index_cast %0 : index to i64
    util.global.store %1, @__hoisted_i64 : i64
    util.return
  }
 ```